### PR TITLE
CI: Download triton dependencies once between all wheel build jobs

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121568
* #121733
* #121268

The LLVM download in particular is slow and is hitting timeouts in CI before the
compiler is even called once. I workaround this by moving the download step into
its own job with a 1 hour timeout and in the other jobs we download from an
artifact instead of the download server.

This reduces strain on the download server and gives us a more generous timeout.

### CODE REVIEW

File: .github/scripts/download_triton_packages.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+
+def get_llvm_package_info_rocm() -> Package:
+    system = platform.system()
+    arch = platform.machine()
+    if arch == "aarch64":
+        arch = "arm64"
+    if system == "Darwin":
+        system_suffix = "apple-darwin"
+        arch = platform.machine()
```
Comment (malfet) [inline .github/scripts/download_triton_packages.py:86]: This is all moot for rocm (as Linux only solution, isn't it?)

File: .github/scripts/download_triton_packages.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        return get_llvm_package_info_cuda()
+
+
+def open_url(url: str) -> Any:
+    user_agent = (
+        "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/119.0"
+    )
+    headers = {
+        "User-Agent": user_agent,
+    }
```
Comment (malfet) [inline .github/scripts/download_triton_packages.py:127]: Why do we need this? What's wrong with curl as user agent?

**Review by malfet — CHANGES_REQUESTED** [review]
Overall change looks fine to me, but imo it makes package upgradess harder, rather than easier, as in addition to updating pin one would generally need to check if download locations have changed, but folks would often forget to do it and as result, this script will be preparing some downloads which are not reused by the underlying triton build system.

As trition pin updates are quite infrequent, I propose we keep it as is

File: .github/scripts/download_triton_packages.py (hunk 1 above)
Comment (peterbell10) [reply .github/scripts/download_triton_packages.py]: I don't think this code is used in CI, no. However, I don't want to make too many changes from the original code in the rocm-triton repo.

File: .github/scripts/download_triton_packages.py (hunk 2 above)
Comment (peterbell10) [reply .github/scripts/download_triton_packages.py]: Not sure, this is how it's written in triton. 
https://github.com/openai/triton/blob/51ecb6f09db45fcc5f216798177f8393b0d92689/python/setup.py#L158

**Comment by peterbell10** [comment]
> As trition pin updates are quite infrequent, I propose we keep it as is

The problem is that the download server isn't 100% reliable and this was causing timeouts in the wheel build, e.g.
https://github.com/pytorch/pytorch/actions/runs/8210910973/job/22460158014

Maybe that was just a fluke and we can get by without these changes most of the time though. I've split the artifact upload error fix off into a separate PR and we can see how that does.

**Pull request closed without merging**